File: cses/01_introductory/2205_gray-code/solution.py
import sys


def main():
    # Đọc toàn bộ input một lần rồi lấy n
    data = sys.stdin.buffer.read().split()
    n = int(data[0])

    total = 1 << n   # tổng số xâu cần in là 2^n
    out = []
    for i in range(total):
        # Công thức Gray code chuẩn: gray(i) = i XOR (i >> 1)
        g = i ^ (i >> 1)
        # Định dạng thành xâu nhị phân đúng n bit (thêm số 0 ở bên trái)
        out.append(format(g, '0{}b'.format(n)))

    # Gom toàn bộ output rồi ghi một lần (fast I/O)
    sys.stdout.write("\n".join(out) + "\n")


main()

File: cses/01_introductory/2205_gray-code/solution.cpp
#include <iostream>
#include <string>

using namespace std;

int main() {
    // Fast I/O: tắt đồng bộ với stdio và bỏ ràng buộc cin/cout
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;

    const int total = 1 << n;   // tổng số xâu cần in là 2^n
    for (int i = 0; i < total; ++i) {
        // Công thức Gray code chuẩn: gray(i) = i XOR (i >> 1)
        const int gray = i ^ (i >> 1);

        // Chuyển gray thành xâu nhị phân đúng n bit (mặc định toàn số 0)
        string code(n, '0');
        for (int bit = 0; bit < n; ++bit) {
            // Nếu bit thứ 'bit' bật thì đặt '1' vào vị trí tương ứng
            if (gray & (1 << bit)) {
                code[n - 1 - bit] = '1';
            }
        }
        cout << code << '\n';
    }

    return 0;
}
